metrics/internal/stackdriver/secret_watcher.go

```diff
@@ -0,0 +1,279 @@
+// NewSecretWatcher constructs a secretWatcher that watches all Secrets.
+func NewSecretWatcher(observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal("", "", observers...)
+}
+
+// NewSecretWatcherSingleNamespace constructs a secretWatcher that watches all Secrets in a namespace.
+func NewSecretWatcherSingleNamespace(namespace string, observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal(namespace, "", observers...)
+}
+
+// NewSecretWatcherSingleSecret constructs a secretWatcher that watches all a single Secret.
+func NewSecretWatcherSingleSecret(namespace string, name string, observers ...Observer) (SecretWatcher, error) {
+	return newSecretWatcherInternal(namespace, name, observers...)
+}
+
+// newSecretWatcherInternal constructs a secretWatcher.
+func newSecretWatcherInternal(namespace string, name string, observers ...Observer) (SecretWatcher, error) {
+	impl := &secretWatcherImpl{
+		SecretNamespace: namespace,
+		SecretName:      name,
+		observers:       observers,
+	}
+
+	if clientErr := impl.setupKubeclient(); clientErr != nil {
+		return nil, clientErr
+	}
+
+	return impl, nil
+}
+
+// secretWatherImpl implmements the secretWatcher interface.
+type secretWatcherImpl struct {
+	// kubeclient is the in-cluster Kubernetes kubeclient, which is lazy-initialized on first use.
+	kubeclient kubernetes.Interface
+
+	// watchLock ensures that only one watch is being run at a time.
+	watchLock sync.Mutex
+	// informer watches the Stackdriver secret if useStackdriverSecretEnabled is true.
+	informer cache.SharedIndexInformer
+	// stopCh stops the secretWatcherInformer when required.
+	stopCh chan struct{}
+	// watchStarted is whether or not the informer is running for the secret being watched.
+	watchStarted bool
+
+	// SecretName is the name of the secret being watched.
+	SecretName string
+	// SecretNamespace is the namespace of the secret being watched.
+	SecretNamespace string
+	// observers is the list of observers to trigger when the Secret is changed.
+	observers []Observer
+}
+
+// StartWatch implements StartWatch of secretWatcher interface.
+func (s *secretWatcherImpl) StartWatch() error {
+	s.watchLock.Lock()
+	defer s.watchLock.Unlock()
+	if s.watchStarted {
+		return nil
+	}
+	s.watchStarted = true
+
+	s.setupInformer()
+	s.stopCh = make(chan struct{})
+	go s.informer.Run(s.stopCh)
+
+	return nil
+}
+
+// StopWatch implements StopWatch secretWatcher interface.
+func (s *secretWatcherImpl) StopWatch() {
+	s.watchLock.Lock()
+	defer s.watchLock.Unlock()
+	if !s.watchStarted {
+		return
+	}
+	s.watchStarted = false
+
+	close(s.stopCh)
+}
+
+// GetSecret implements GetSecret of secretWatcher interface.
+func (s *secretWatcherImpl) GetSecret(namespace string, name string) (*corev1.Secret, error) {
+	if namespace == "" || name == "" {
+		return nil, fmt.Errorf("Must specify a non-empty Secret namespace and name to get, namespace specified: %v, name specified: %v", namespace, name)
+	}
```

Comment:
Should this check also live on the NewSecretWatcher initializers?

Reply:
Yes, that would be a good check, I'll add it.